pkg/controller/storagecluster/initialization_reconciler.go

```diff
@@ -55,12 +80,140 @@ func (r *ReconcileStorageCluster) ensureStorageClasses(instance *ocsv1.StorageCl
+			Name:      externalClusterDetailsSecret,
+			Namespace: instance.Namespace,
+		},
+	}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: found.Name, Namespace: found.Namespace}, found)
+	if err != nil {
+		return err
+	}
+	ownerRef := metav1.OwnerReference{
+		UID:        instance.UID,
+		APIVersion: instance.APIVersion,
+		Kind:       instance.Kind,
+		Name:       instance.Name,
+	}
+	var data []ExternalResource
+	err = json.Unmarshal(found.Data[externalClusterDetailsKey], &data)
+	if err != nil {
+		reqLogger.Error(err, "could not parse json blob")
+		return err
+	}
+	scs, err := r.newStorageClasses(instance)
+	if err != nil {
+		return err
+	}
+	// newStorageClass might not return OBC storageClass so checking the
+	// number of storageClasses returned
+	if len(scs) < 3 {
+		scs = append(scs, r.newOBCStorageClass(instance))
+	}
+	for _, d := range data {
+		objectMeta := metav1.ObjectMeta{
+			Name:            d.Name,
+			Namespace:       instance.Namespace,
+			OwnerReferences: []metav1.OwnerReference{ownerRef},
+		}
+		objectKey := types.NamespacedName{Name: d.Name, Namespace: instance.Namespace}
+		switch d.Kind {
+		case "ConfigMap":
+			cm := &corev1.ConfigMap{
+				ObjectMeta: objectMeta,
+				Data:       d.Data,
+			}
+			found := &corev1.ConfigMap{ObjectMeta: objectMeta}
+			err := r.client.Get(context.TODO(), objectKey, found)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					reqLogger.Info(fmt.Sprintf("creating configmap: %s", cm.Name))
+					err = r.client.Create(context.TODO(), cm)
+					if err != nil {
+						reqLogger.Error(err, "creation of configmap failed")
+						return err
+					}
+				} else {
+					reqLogger.Error(err, "Unable the get the configmap")
+					return err
+				}
+			}
+		case "Secret":
+			sec := &corev1.Secret{
+				ObjectMeta: objectMeta,
+				Data:       make(map[string][]byte),
+			}
+			for k, v := range d.Data {
+				sec.Data[k] = []byte(v)
+			}
+			found := &corev1.Secret{ObjectMeta: objectMeta}
+			err := r.client.Get(context.TODO(), objectKey, found)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					reqLogger.Info(fmt.Sprintf("creating secret: %s", sec.Name))
+					err = r.client.Create(context.TODO(), sec)
+					if err != nil {
+						reqLogger.Error(err, "creation of secret failed")
+						return err
+					}
+				} else {
+					reqLogger.Error(err, "unable the get the secret")
+					return err
+				}
+			}
+		case "StorageClass":
+			var sc *storagev1.StorageClass
+			if d.Name == cephfs {
+				// Setting the fsname for cephfs StorageClass
+				sc = scs[0]
+			} else if d.Name == cephRBD {
+				// Setting the PoolName for RBD StorageClass
+				sc = scs[1]
+			} else if d.Name == cephRGW {
+				// Setting the Endpoint for OBC StorageClass
+				sc = scs[2]
+			}
+			for k, v := range d.Data {
+				sc.Parameters[k] = v
+			}
+		}
+	}
+	// creating all the storageClass once we set the values
+	err = r.createStorageClasses(scs, reqLogger)
+	if err != nil {
+		return err
```

Comment:
name your error

Reply:
`createStorageClasses()` should be responsible for elaborating on its own errors, we just need to pass it up. This is outside the scope of this PR.